consensus/fork.go

```diff
@@ -4,119 +4,115 @@ import (
 	"github.com/NebulousLabs/Sia/crypto"
 )
 
-// backtrackToBlockchain returns a list of nodes that go from the current node
-// to the first common parent. The common parent will be the final node in the
-// slice.
-func (s *State) backtrackToBlockchain(bn *blockNode) (nodes []*blockNode) {
-	nodes = append(nodes, bn)
+// invalidateNode recursively moves the children of a blockNode to the bad
+// blocks list.
+func (s *State) invalidateNode(node *blockNode) {
+	for i := range node.children {
+		s.invalidateNode(node.children[i])
+	}
+
+	delete(s.blockMap, node.block.ID())
+	s.badBlocks[node.block.ID()] = struct{}{}
+}
+
+// backtrackToBlockchain traces backwards from 'bn' until it reaches a node in
+// the State's current path (the "common parent"). It returns the (inclusive)
+// set of nodes between the common parent and 'bn', starting from the former.
+func (s *State) backtrackToBlockchain(bn *blockNode) []*blockNode {
+	path := []*blockNode{bn}
 	for s.currentPath[bn.height] != bn.block.ID() {
 		bn = bn.parent
-		nodes = append(nodes, bn)
+		// prepend, not append
+		path = append([]*blockNode{bn}, path...)
 
 		// Sanity check - all block nodes should have a parent except the
-		// genesis block. This loop should break before reaching the genesis
-		// block.
+		// genesis block, and this loop should break before reaching the
+		// genesis block.
 		if bn == nil {
 			if DEBUG {
 				panic("backtrack hit a nil node?")
 			}
-			return
+			break
 		}
 	}
-	return
+	return path
 }
 
-// rewindToNode will rewind blocks until `bn` is the highest block, returning
-// the list of nodes that got rewound.
-func (s *State) rewindToNode(bn *blockNode) (rewoundNodes []*blockNode) {
+// rewindToNode will rewind blocks from the State's current path until 'bn' is
+// the current block.
+func (s *State) rewindToNode(bn *blockNode) {
 	// Sanity check - make sure that bn is in the currentPath.
 	if DEBUG {
-		if bn.block.ID() != s.currentPath[bn.height] {
-			panic("bad use of rewindToNode")
+		if s.currentPath[bn.height] != bn.block.ID() {
```

Comment:
I tend to write these comparisons as I would say them out loud. "If the State doesn't match the blockNode" as opposed to "If the blockNode doesn't match the State". This particular instance is pretty ambiguous though; I don't think it was worth reordering.